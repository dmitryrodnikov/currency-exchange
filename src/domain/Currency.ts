export enum Currency {
    USD = 'USD',
    EUR = 'EUR',
    GBP = 'GBP',
}

export const Currencies = Object.keys(Currency);
