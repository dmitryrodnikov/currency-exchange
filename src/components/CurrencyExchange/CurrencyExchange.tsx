import * as React from 'react';
import styled from 'styled-components';
import {Input, InputChangeHandler} from '../Input/Input';
import {CurrencySelect} from '../CurrencySelect/CurrencySelect';
import {Currency, CurrencySymbol} from '../../domain/Currency';
import {ButtonClickHandler} from '../Button/Button';
import {Output} from '../Output/Output';

interface CurrencyExchangeProps {
    fromBalance: string;
    fromCurrency: Currency;
    fromAmount: string;
    toBalance: string;
    toCurrency: Currency;
    toAmount: string;
    toExchangeRate: string;
    onChangeAmount: InputChangeHandler,
    onChangeCurrencyFrom: ButtonClickHandler<Currency>;
    onChangeCurrencyTo: ButtonClickHandler<Currency>;
    onClickExchangeButton: VoidFunction;
    currencyList: Currency[]
}

const StyledCurrencyExchange = styled.div`
    display: flex;
    flex-direction: column;
    padding: 18px 12px;
    border-radius: 12px;
    box-shadow: 0 2px 4px #00000020, 0 4px 12px #00000010;
`;

const ExchangeBlockWrapper = styled.div`
    padding: 16px 12px;
    overflow: hidden;
    
    &:first-child {
        border-bottom: 2px solid #00000010;
    }
`;

const InfoBlock = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-between;
    overflow: hidden;
    color: #00000080;
    font-size: 14px;
    margin: 6px 8px 0 0;
`;

const ExchangeBlock = styled.div`
    display: flex;
    align-items: center;
    overflow: hidden;
`;

const CurrencyAmount = styled.div`
    margin-left: 20px;
    width: 100%;
    overflow: hidden;
`;

const ExchangeButtonWrapper = styled.div`
    display: flex;
    justify-content: flex-end;
`;

const ExchangeButton = styled.div`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: 8px 12px;
    color: #fff;
    background: #1200ff;
    border-radius: 8px;
    cursor: pointer;
`;

export function CurrencyExchange(
    {
        fromBalance,
        fromCurrency,
        fromAmount,
        toBalance,
        toCurrency,
        toAmount,
        toExchangeRate,
        currencyList,
        onChangeAmount,
        onChangeCurrencyFrom,
        onChangeCurrencyTo,
        onClickExchangeButton,
    }: CurrencyExchangeProps
) {
    const toExchangeRateFormatted = `${CurrencySymbol[toCurrency]}1 = ${CurrencySymbol[fromCurrency]}${toExchangeRate}`;
    const getFormattedBalance = (amount: string, currency: Currency) => `You have ${CurrencySymbol[currency]}${amount}`;
    const toBalanceFormatted = getFormattedBalance(toBalance, toCurrency);
    const fromBalanceFormatted = getFormattedBalance(fromBalance, fromCurrency);
    return (
        <StyledCurrencyExchange>
            <ExchangeBlockWrapper>
                <ExchangeBlock>
                    <CurrencySelect
                        selected={fromCurrency}
                        currencies={currencyList}
                        onItemClick={onChangeCurrencyFrom}
                    />
                    <CurrencyAmount>
                        <Input
                            value={fromAmount}
                            onChange={onChangeAmount}
                        />
                    </CurrencyAmount>
                </ExchangeBlock>
                <InfoBlock>
                    <div>{fromBalanceFormatted}</div>
                </InfoBlock>
            </ExchangeBlockWrapper>
            <ExchangeBlockWrapper>
                <ExchangeBlock>
                    <CurrencySelect
                        selected={toCurrency}
                        currencies={currencyList}
                        onItemClick={onChangeCurrencyTo}
                    />
                    <CurrencyAmount>
                        <Output text={toAmount}/>
                    </CurrencyAmount>
                </ExchangeBlock>
                <InfoBlock>
                    <div>{toBalanceFormatted}</div>
                    <div>{toExchangeRateFormatted}</div>
                </InfoBlock>
            </ExchangeBlockWrapper>
            <ExchangeButtonWrapper>
                <ExchangeButton onClick={onClickExchangeButton}>
                    Exchange
                </ExchangeButton>
            </ExchangeButtonWrapper>
        </StyledCurrencyExchange>
    );
}
