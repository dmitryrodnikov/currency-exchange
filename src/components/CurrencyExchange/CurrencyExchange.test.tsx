import React from 'react';
import {render, shallow} from 'enzyme';
import {CurrencyExchange} from './CurrencyExchange';
import {Currency} from '../../domain/Currency';

const CURRENCY_LIST= [Currency.USD, Currency.GBP, Currency.EUR];

describe('CurrencyExchange', () => {
    it('Shallow render snapshots are equal', () => {
        const wrapper = shallow(
            <CurrencyExchange
                fromCurrency={Currency.USD}
                toCurrency={Currency.EUR}
                fromAmount={String(100)}
                toAmount={String(80)}
                onChangeAmount={() => {}}
                currencyList={CURRENCY_LIST}
                onChangeCurrencyFrom={() => {}}
                onChangeCurrencyTo={() => {}}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('Render snapshots are equal', () => {
        const wrapper = render(
            <CurrencyExchange
                fromCurrency={Currency.USD}
                toCurrency={Currency.EUR}
                fromAmount={String(100)}
                toAmount={String(80)}
                onChangeAmount={() => {}}
                currencyList={CURRENCY_LIST}
                onChangeCurrencyFrom={() => {}}
                onChangeCurrencyTo={() => {}}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
