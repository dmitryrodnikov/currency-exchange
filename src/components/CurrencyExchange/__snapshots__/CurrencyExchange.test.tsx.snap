// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyExchange Render snapshots are equal 1`] = `
<div
  class="sc-ifAKCX bcPcvD"
>
  <div
    class="sc-EHOje RXNYV"
  >
    <div
      class="sc-gzVnrw jnOtFG"
    >
      <div
        class="sc-htpNat gDqwhI"
      >
        <div
          class="sc-bwzfXH jsrFJT"
        >
          USD
        </div>
        <div
          class="sc-bwzfXH bRAeMy"
        >
          GBP
        </div>
        <div
          class="sc-bwzfXH bRAeMy"
        >
          EUR
        </div>
      </div>
      <div
        class="sc-htoDjs gktCDC"
      >
        <input
          class="sc-bdVaJa jSDdor"
          type="text"
          value="100"
        />
      </div>
    </div>
    <div
      class="sc-bZQynM jnNfil"
    >
      <div>
        You have $undefined
      </div>
    </div>
  </div>
  <div
    class="sc-EHOje RXNYV"
  >
    <div
      class="sc-gzVnrw jnOtFG"
    >
      <div
        class="sc-htpNat gDqwhI"
      >
        <div
          class="sc-bwzfXH bRAeMy"
        >
          USD
        </div>
        <div
          class="sc-bwzfXH bRAeMy"
        >
          GBP
        </div>
        <div
          class="sc-bwzfXH jsrFJT"
        >
          EUR
        </div>
      </div>
      <div
        class="sc-htoDjs gktCDC"
      >
        <div
          class="sc-bxivhb hIASvP"
        >
          80
        </div>
      </div>
    </div>
    <div
      class="sc-bZQynM jnNfil"
    >
      <div>
        You have €undefined
      </div>
      <div>
        €1 = $undefined
      </div>
    </div>
  </div>
</div>
`;

exports[`CurrencyExchange Shallow render snapshots are equal 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <CurrencySelect
        currencies={
          Array [
            "USD",
            "GBP",
            "EUR",
          ]
        }
        onItemClick={[Function]}
        selected="USD"
      />
      <styled.div>
        <Input
          onChange={[Function]}
          value="100"
        />
      </styled.div>
    </styled.div>
    <styled.div>
      <div>
        You have $undefined
      </div>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <CurrencySelect
        currencies={
          Array [
            "USD",
            "GBP",
            "EUR",
          ]
        }
        onItemClick={[Function]}
        selected="EUR"
      />
      <styled.div>
        <Output
          text="80"
        />
      </styled.div>
    </styled.div>
    <styled.div>
      <div>
        You have €undefined
      </div>
      <div>
        €1 = $undefined
      </div>
    </styled.div>
  </styled.div>
</styled.div>
`;
